pkg/apis/clusteroperator/types.go

```diff
@@ -178,15 +185,51 @@ type AWSRegionAMIs struct {
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
-// ClusterProviderConfigSpec is the cluster specification stored in the
-// ProviderConfig of a cluster.k8s.io Cluster.
-type ClusterProviderConfigSpec struct {
+// AWSClusterProviderConfig is the AWS specific cluster specification stored in the ProviderConfig of an AWS cluster.k8s.io.Cluster.
+type AWSClusterProviderConfig struct {
 	// +optional
 	metav1.TypeMeta
 	// +optional
 	metav1.ObjectMeta
 
-	ClusterDeploymentSpec
+	ClusterProviderConfig
+
+	// AWS specifies cluster hardware configuration on AWS.
+	Hardware AWSClusterSpec
+
+	// DefaultHardwareSpec specifies the default AWS hardware spec for an AWS machine.
+	// +optional
+	DefaultHardware *MachineSetAWSHardwareSpec
+
+	// OpenShiftConfig contains configuration specific to the OpenShift distribution.
+	OpenShiftConfig OpenShiftDistributionConfig
+}
+
+// ClusterProviderConfig contains cluster specification relevant to all cloud providers.
+type ClusterProviderConfig struct {
+	// ClusterID is the ID to assign to the cluster in the cloud provider.
+	// This will be generated by the API Server when the ClusterDeployment is
+	// created. It should not be set by the user.
+	ClusterID string
+}
+
+// OpenShiftDistributionConfig contains configuration specific to the OpenShift distribution.
+type OpenShiftDistributionConfig struct {
+	ClusterConfigSpec
+
+	// Version contains information on the OpenShift version that should be installed on this cluster.
+	Version OpenShiftDistributionConfigVersion
```

Comment:
Should this also be a member of ClusterVersionSpec as well ? (instead of duplicating the fields there)

Reply:
The objects are starting to look weird no matter how I slice this.

We could do:

```
ClusterVersion
- VMImages
- OpenShiftConfigVersion
  - DeploymentType
  - Version
  - Images
OpenShiftDistributionConfig
- OpenShiftConfigVersion
  - DeploymentType
  - Version
  - Images
- other things
```

Or I could move Images up one level on both.

Reply:
true... thinking about it some more, I'd be ok leaving it as is.